src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java

```diff
@@ -0,0 +1,246 @@
+     * its content will be used as initial content for this collection. Otherwise it is created empty.
+     *
+     * @param id the id of the Couchbase document to back the set.
+     * @param bucket the {@link Bucket} through which to interact with the document.
+     */
+    public CouchbaseArraySet(String id, Bucket bucket) {
+        this.id = id;
+        this.bucket = bucket;
+
+        try {
+            this.bucket.insert(JsonArrayDocument.create(id));
+        } catch (DocumentAlreadyExistsException e) {
+            //use a pre-existing document
+        }
+    }
+
+    /**
+     * Create a new {@link CouchbaseArraySet}, backed by the document identified by <code>id</code>
+     * in the given Couchbase <code>bucket</code>. Note that if the document already exists, its content is
+     * reset to the values copied from the given <code>data</code> Map.
+     *
+     * A null or empty data map will re-initialize any pre-existing document to an empty content.
+     *
+     * @param id the id of the Couchbase document to back the set.
+     * @param bucket the {@link Bucket} through which to interact with the document.
+     * @param initialData Set of the elements to initially store in the CouchbaseArraySet.
+     */
+    public CouchbaseArraySet(String id, Bucket bucket, Set<? extends T> initialData) {
+        this.id = id;
+        this.bucket = bucket;
+
+        JsonArray data = JsonArray.create();
+        if (initialData != null && !initialData.isEmpty()) {
+            for (Object o : data) {
+                data.add(o);
+            }
+        }
+        bucket.upsert(JsonArrayDocument.create(id, data));
+    }
+
+    @Override
+    public int size() {
+        //TODO use subdoc GET_COUNT when available
+        JsonArrayDocument current = bucket.get(id, JsonArrayDocument.class);
+        return current.content().size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        DocumentFragment<Lookup> current = bucket.lookupIn(id).exists("[0]").execute();
+        return current.status(0) == ResponseStatus.SUBDOC_PATH_NOT_FOUND;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        //TODO subpar implementation for a Set, use ARRAY_CONTAINS
+        JsonArrayDocument current = bucket.get(id, JsonArrayDocument.class);
+        for (Object in : current.content()) {
+            if (in.equals(o)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        JsonArrayDocument current = bucket.get(id, JsonArrayDocument.class);
+        return new Itr(current);
+    }
+
+    @Override
+    public boolean add(T t) {
+        if (t == null) {
+            throw new NullPointerException("Unsupported null value");
+        }
+
+        for (int i = 0; i < MAX_OPTIMISTIC_LOCKING_ATTEMPTS; i++) {
+            try {
+                JsonArrayDocument current = bucket.get(id, JsonArrayDocument.class);
```

Comment:
it is not exactly like contains as we have to extract the cas and take it into account for the subsequent append. That said, maybe I could use ARRAY_ADD_UNIQUE... but then it has an even greater limitation (cannot accept anything that is non-primitive, like `JsonObject` or `JsonArray`).